lectures/olg.md

```diff
@@ -46,114 +47,852 @@ The OLG model takes up this challenge.
+    g(k^*) = k^*  \left [ 1 + \beta^{-1/\gamma} (\alpha (k^*)^{\alpha-1})^{(\gamma-1)/\gamma} \right ] - (1-\alpha)(k^*)^{\alpha}
+```
+
+```{code-cell} ipython3
+def g(k_star, model):
+    α, β, γ = model.α, model.β, model.u.__defaults__[0]
+    z = (1 - α) * k_star**α
+    R1 = α ** (1-1/γ)
+    R2 = k_star**((α * γ - α + 1) / γ)
+    p = k_star + k_star * β**(-1/γ) * R1 * R2
+    return p - z
+```
+
+```{code-cell} ipython3
+k_star = optimize.newton(g, 0.2, args=(m_crra,))
+print(f"k_star = {k_star}")
+```
+
+```{code-cell} ipython3
+plot_45(m_crra, k_next, k_star)
 ```
 
 
 
-A steady state is given by a solution to this equation such that
-$k_{t+1} = k_t = k^*$, i.e,
+The next figure shows three time paths for capital, from
+three distinct initial conditions, under the parameterization listed above.
+
+At this parameterization, $k^* \approx 0.161$.
+
+Let's define the constants and three distinct intital conditions
+
+```{code-cell} ipython3
+ts_length = 10
+x0 = np.array([0.001, 1.2, 2.6])
+```
+
+```{code-cell} ipython3
+def simulate_ts(m, x0_values, ts_length):
+
+    k_star = optimize.newton(g, 0.2, args=(m,))
+    fig, ax = plt.subplots(figsize=(10, 5))
+
+    ts = np.zeros(ts_length)
+
+    # simulate and plot time series
+    for x_init in x0_values:
+        ts[0] = x_init
+        for t in range(1, ts_length):
+            ts[t] = k_next(ts[t-1], m)
+        ax.plot(np.arange(ts_length), ts, '-o', ms=4, alpha=0.6,
+                label=r'$k_0=%g$' %x_init)
+    ax.plot(np.arange(ts_length), np.full(ts_length,k_star),
+            alpha=0.6, color='red', label=r'$k^*$')
+    ax.legend(fontsize=10)
+
+    ax.set_xlabel(r'$t$', fontsize=14)
+    ax.set_ylabel(r'$k_t$', fontsize=14)
+
+    plt.show()
+```
+
+```{code-cell} ipython3
+simulate_ts(m_crra, x0, ts_length)
+```
+
+## Exercises
+
+TODO Add exercise-solution environment
+
+
+
+### Exercise 1
+
+To solve the equilibrium of the CRRA case numerically using [](equilibrium_crra).
+
+Plot the equilibrium quantity and equilibrium price in the equilibrium plot with the CRRA utility (TODO label and refer the plot generated by the code).
+
+
+
+### Solution to exercise 1
+
+To solve the equation we need to turn to Newton's method.
+
+
 
 ```{math}
-:label: k_star
-    k^* = \frac{s(f(k^*)-k^*f'(k^*), f'(k^*))}{1+n}
+    L (1-\alpha)(K_t / L)^{\alpha} \left [ 1 + \beta^{-1/\gamma} R_{t+1}^{(\gamma-1)/\gamma} \right ]^{-1} = L\left (\frac{R_{t+1}}{\alpha} \right )^{1/(\alpha - 1)}
+```
+
+```{code-cell} ipython3
+def find_Rstar_newton(x, K_prev, model):
+    α, β, γ, L = model.α, model.β, model.u.__defaults__[0], model.L
+    lhs = L * (1-α) * (K_prev / L)**α
+    lhs /= (1 + β**(-1/γ) * x**((γ-1)/γ))
+    # To should always non-negative for a real solution.
+    assert x>=0, x
+    rhs = L * (x / α)**(1/(α-1))
+    return lhs - rhs
+```
+
+```{code-cell} ipython3
+def find_Rstar(K_prev, model):
+    return optimize.newton(find_Rstar_newton, K_prev*100, args=(K_prev, model))
+
```

Comment:
This is the code for finding $R^*$ in the CRRA case. Seems like the newton method is unable to find the solution because while iterating it goes to a value where `x<0` and the function returns complex values as $\alpha-1$ is also less than 0. And so the CI is failing.

Reply:
I fixed this issue by plotting `find_Rstar_newton` for some values and observed the root in the plot. Keeping the init value in `optimize.newton` closer to the observed value fixes the problem.